### PR TITLE
RBAC: write enforced rules results to dynamic metadata

### PR DESCRIPTION
**Pull request by antoniovleonti**
Commit Message: RBAC: write enforced rules results to dynamic metadata
Additional Description: adding enforced rules stats to dynamic metadata (like shadow stats are already done) so they will be available later in access logs filters.
Risk Level: low
Testing: unit tested
Docs Changes: none
Release Notes: none
Platform Specific Features: none

### CODE REVIEW

**Review by rshriram — APPROVED** [review]

File: source/extensions/filters/http/rbac/rbac_filter.cc [hunk 1]
```diff
@@ -190,9 +199,15 @@ RoleBasedAccessControlFilter::decodeHeaders(Http::RequestHeaderMap& headers, boo
...
       }
+
+      *fields[config_->enforcedEngineResultField()].mutable_string_value() =
+          Filters::Common::RBAC::DynamicMetadataKeysSingleton::get().EngineResultDenied;
+      callbacks_->streamInfo().setDynamicMetadata("envoy.filters.http.rbac", metrics);
+
       return Http::FilterHeadersStatus::StopIteration;
     }
   }
+  callbacks_->streamInfo().setDynamicMetadata("envoy.filters.http.rbac", metrics);
```
Comment (adisuissa) [inline source/extensions/filters/http/rbac/rbac_filter.cc:210]: This should probably only be called if `metrics`/`fields` isn't empty.

File: source/extensions/filters/http/rbac/rbac_filter.h [hunk 2]
```diff
@@ -66,6 +66,15 @@ class RoleBasedAccessControlFilterConfig {
            Filters::Common::RBAC::DynamicMetadataKeysSingleton::get().ShadowEngineResultField;
   }
 
+  std::string enforcedEffectivePolicyIdField() const {
+    return rules_stat_prefix_ + Filters::Common::RBAC::DynamicMetadataKeysSingleton::get()
```
Comment (adisuissa) [inline source/extensions/filters/http/rbac/rbac_filter.h:70]: I don't really understand why are these considered stats (as they are prefixed by `rules_stat_prefix_`).
Seems like an error in the original shadow fields, an it is being persisted here as well.

I suggest considering not using a prefix at all, but will leave this to the code-owner to decide.
cc @yangminzhu @yanavlasov

**Review by adisuissa — COMMENTED** [review]
Thanks!
Please add a release note.
Left a comment that a code-owner and senior-maintainer should decide upon.
/assign @yanavlasov

File: source/extensions/filters/http/rbac/rbac_filter.h (hunk 2 above)
Comment (antoniovleonti) [reply source/extensions/filters/http/rbac/rbac_filter.h:70]: I think the prefix is useful when there are multiple RBAC filters configured.

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**